packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -151,3 +157,319 @@ export class Intc extends Op {
+export class txn extends Op {
+  execute (stack: TEALStack): void {
+    // txn CloseRemainderTo
+    // push closeRemainderTo from transaction data
+  }
+}
+
+export class Pop extends Op {
+  execute (stack: TEALStack): void {
+    stack.pop();
+  }
+}
+
+// duplicate last value on stack
+export class Dup extends Op {
+  execute (stack: TEALStack): void {
+    const lastValue = stack.pop() as bigint;
+   
+    stack.push(lastValue);
+    stack.push(lastValue);
+  }
+}
+
+// duplicate two last values on stack: A, B -> A, B, A, B
+export class Dup2 extends Op {
+  execute (stack: TEALStack): void {
+    const lastValueA = stack.pop() as bigint;
+    const lastValueB = stack.pop() as bigint;
+    
+    stack.push(lastValueA);
+    stack.push(lastValueB);
+    stack.push(lastValueA);
+    stack.push(lastValueB);
+  }
+}
+
+// pop two byte strings A and B and join them, push the result
+// concat panics if the result would be greater than 4096 bytes.
+export class Concat extends Op {
+  execute (stack: TEALStack): void {
+    const valueB = stack.pop();
+    const valueA = stack.pop();
+
+    if (typeof valueA === "bigint" || typeof valueB === "bigint") {
+      throw new TealError(ERRORS.TEAL.INVALID_TYPE);
+    } else if (typeof valueA === "undefined" || typeof valueB === "undefined") {
+      throw new TealError(ERRORS.TEAL.INVALID_TYPE);
+    } else {
+      if (valueA.length + valueB.length > MaxStringSize) {
+        throw new TealError(ERRORS.TEAL.CONCAT_ERROR);
+      }
+      var c = new Uint8Array(valueB.length + valueA.length);
+      c.set(valueB);
+      c.set(valueA, valueB.length);
+      stack.push(c);
+    }
+  }
+}
+
+// pop a byte string X. For immediate values in 0..255 M and N:
+// extract a range of bytes from it starting at M up to but not including N,
+// push the substring result. If N < M, or either is larger than the string length,
+// the program fails
+export class Substring extends Op {
+  execute (stack: TEALStack): void {
+    // const byteString = stack.pop();
+  }
+}
+
+// pop a byte string A and two integers B and C.
+// Extract a range of bytes from A starting at B up to
+// but not including C, push the substring result. If C < B,
+// or either is larger than the string length, the program fails
+export class Substring3 extends Op {
+  execute (stack: TEALStack): void {
+    const byteString = stack.pop();
+    const start = stack.pop();
+    const end = stack.pop();
+
+    if (typeof byteString === "undefined" || typeof byteString === "bigint") {
+      throw new TealError(ERRORS.TEAL.INVALID_TYPE);
+    }
+
+    if (typeof start === "bigint" && typeof end === "bigint") {
+      if (end < start) {
+        throw new Error("substring end before start");
+      }
+      if (start > byteString.length || end > byteString.length) {
+        throw new Error("substring range beyond length of string");
+      }
+
+      stack.push(byteString.slice(Number(start), Number(end)));
+    } else {
```

Comment:
- Check stack size (>=3)
- assert 1st value (`byteString`) as `bytes`
- assert `start` and `end` as `BigInt`.
- remove lines - 458, 462 checks
- Errors should be `TealErrors`